### PR TITLE
added jupyter notebooks for discretisation, closes #60

### LINKED ISSUE #60: improve discretisers jupyter demos
Separate each discretiser into a single jupyter notebook. 

Check if new functionality has been added and add an example in the notebook.

### PR DESCRIPTION
**Pull request by ashok49473**
Added jupyter notebooks for discretisation transformers.

### CODE REVIEW

**Review by solegalli — CHANGES_REQUESTED** [review]
Hi @ashok49473 

thanks a lot for the PR with the changes. This is looking really good.  

I have a couple of comments / suggestions:

- I dont think we need `%matplotlib inline `with the latest versions of jupyter notebook. If you remove it does the notebook not run as well? 

**ArbitraryDiscretiser.ipynb**
-Could we choose different interval limits so that all the bins show in the plots? In the current version we have limits for 6 intervals, but then only 3 come in the plots. 

- Could we add a bar plot also for when the transformer returns the boundaries? Or do you think it is an over-kill?

- It would be great if the intervals for each variable were more obviously different. At the moment they are very similar, and one of the variables is very skewed so i think it would help to create more intervals towards the values where the observations accumulate.

**DecisionTreeDiscretiser.ipynb**

- Could we include a plot of the bins vs the mean of the target for each bin? Tree encoding should in theory return a monotonic relationship, so maybe we can highlight this to new users.

**EqualFrequencyDiscretiser.ipynb**
- all good

**EqualWidthDiscretiser.ipynb**

- could we, after the cell that shows the boundaries, highlight that the intervals are more or less of the same length? it would be by taking the difference between the limits :  
```
nterval(768.2, 1202.4, closed='right'),
       Interval(1202.4, 1636.6, closed='right'),
       Interval(1636.6, 2070.8, closed='right'),
       Interval(2070.8, 2505.0, closed='right'),
       Interval(2505.0, 2939.2, closed='right'),
       Interval(2939.2, 3373.4, closed='right'),
       Interval(4241.8, inf, closed='right')], dtype=object)
```
would be 1202.4 - 768.2,  for the first interval, and the difference should be similar for all intervals.

**New notebooks":
the discretisers equalwidth, equalfrequency and arbitrary have a parameter called `return_object`, where the user can returns the bin numbers as `type object` instead of the default type numeric.

This is so that they can use encoding at the back of discretisation. Could you create 3 additional notebooks where you couple for example:
- equalfrequency + WoE
- equalwidth + ordered ordinal
- arbitrary + MeanEncoder?

Could you do the 2 transformations in single steps and then later on in the notebook inside a pipeline and also, show how the bins are ordered monotonically with the target after applying the encoder?

This is very useful for linear models, because by using discretisation + a monotonic encoding we create monotonic variables, from those that before were not originally. And this tends to hep improve the performance of the model. And maybe explain all of these int the new notebooks?

**Encoders:**

Did you make additional changes to the encoders? because they were merged to master in the previous PR, so in order not to carry the changes to this PR, you need to pull from upstream master and merge to your fork master. And then create a new feature branch from the updated master. Alternatively, you need to git pull rebase master into your feature branch. More details here:
https://feature-engine.readthedocs.io/en/latest/contributing/github_quick_start.html
Check for **"keep your fork updated"** and **"what happens after a PR is merged"**

In summary: we should not have the encoder notebooks in this PR. Could you somehow remove them or make a new PR just with the discretisers?

I also included 2 additional comments for the .py files

Thanks a lot!

File: feature_engine/encoding/decision_tree.py [hunk 1]
```diff
@@ -14,11 +14,11 @@
 
 class DecisionTreeEncoder(BaseCategoricalTransformer):
```
[the remark this answers is not in the document]
Comment (ashok49473) [reply feature_engine/encoding/decision_tree.py:15]: Hi @solegalli mam,
Actually I'm new to open source contribution and github. I will remove all the encoders notebooks from this PR.
I will start working on new changes in this PR.
Thanks for your guidance.

[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/encoding/decision_tree.py:15]: Hey @ashok49473 

No problem at all. We are all learning, myself included :) 
Cheers

File: feature_engine/discretisation/arbitrary.py [hunk 2]
```diff
@@ -10,14 +10,14 @@
 
 class ArbitraryDiscretiser(BaseNumericalTransformer):
```
[the remark this answers is not in the document]
Comment (ashok49473) [reply feature_engine/discretisation/arbitrary.py:11]: I will remove it

**Comment by ashok49473** [comment]
Hello @solegalli mam,
I got the below error when I tried to fit DecisionTreeDiscretiser on iris dataset. 
ValueError: multiclass format is not supported

**Comment by solegalli** [comment]
Hello @ashok49473 Could you either attach a copy / screenshot of the error, or upload the notebooks so I can have a look at where this might be coming from?

**Comment by ashok49473** [comment]
**unable to upoad the ipython notebook ,so pasted the error **
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-54-381d35a26b12> in <module>
      4                                    regression=False)
      5 
----> 6 treeDisc.fit(X_train, y_train)

c:\users\king_ashok\desktop\feature_engine\feature_engine\discretisation\decision_tree.py in fit(self, X, y)
    140 
    141             # fit the model to the variable
--> 142             tree_model.fit(X[var].to_frame(), y)
    143 
    144             self.binner_dict_[var] = tree_model

~\anaconda3\lib\site-packages\sklearn\utils\validation.py in inner_f(*args, **kwargs)
     71                           FutureWarning)
     72         kwargs.update({k: arg for k, arg in zip(sig.parameters, args)})
---> 73         return f(**kwargs)
     74     return inner_f
     75 

~\anaconda3\lib\site-packages\sklearn\model_selection\_search.py in fit(self, X, y, groups, **fit_params)
    734                 return results
    735 
--> 736             self._run_search(evaluate_candidates)
    737 
    738         # For multi-metric evaluation, store the best_index_, best_params_ and

~\anaconda3\lib\site-packages\sklearn\model_selection\_search.py in _run_search(self, evaluate_candidates)
   1186     def _run_search(self, evaluate_candidates):
   1187         """Search all candidates in param_grid"""
-> 1188         evaluate_candidates(ParameterGrid(self.param_grid))
   1189 
   1190 

~\anaconda3\lib\site-packages\sklearn\model_selection\_search.py in evaluate_candidates(candidate_params)
    713                                for parameters, (train, test)
    714                                in product(candidate_params,
--> 715                                           cv.split(X, y, groups)))
    716 
    717                 if len(out) < 1:

~\anaconda3\lib\site-packages\joblib\parallel.py in __call__(self, iterable)
   1002             # remaining jobs.
   1003             self._iterating = False
-> 1004             if self.dispatch_one_batch(iterator):
   1005                 self._iterating = self._original_iterator is not None
   1006 

~\anaconda3\lib\site-packages\joblib\parallel.py in dispatch_one_batch(self, iterator)
    833                 return False
    834             else:
--> 835                 self._dispatch(tasks)
    836                 return True
    837 

~\anaconda3\lib\site-packages\joblib\parallel.py in _dispatch(self, batch)
    752         with self._lock:
    753             job_idx = len(self._jobs)
--> 754             job = self._backend.apply_async(batch, callback=cb)
    755             # A job can complete so quickly than its callback is
    756             # called before we get here, causing self._jobs to

~\anaconda3\lib\site-packages\joblib\_parallel_backends.py in apply_async(self, func, callback)
    207     def apply_async(self, func, callback=None):
    208         """Schedule a func to be run"""
--> 209         result = ImmediateResult(func)
    210         if callback:
    211             callback(result)

~\anaconda3\lib\site-packages\joblib\_parallel_backends.py in __init__(self, batch)
    588         # Don't delay the application, to avoid keeping the input
    589         # arguments in memory
--> 590         self.results = batch()
    591 
    592     def get(self):

~\anaconda3\lib\site-packages\joblib\parallel.py in __call__(self)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~\anaconda3\lib\site-packages\joblib\parallel.py in <listcomp>(.0)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py in _fit_and_score(estimator, X, y, scorer, train, test, verbose, parameters, fit_params, return_train_score, return_parameters, return_n_test_samples, return_times, return_estimator, error_score)
    558     else:
    559         fit_time = time.time() - start_time
--> 560         test_scores = _score(estimator, X_test, y_test, scorer)
    561         score_time = time.time() - start_time - fit_time
    562         if return_train_score:

~\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py in _score(estimator, X_test, y_test, scorer)
    605         scores = scorer(estimator, X_test)
    606     else:
--> 607         scores = scorer(estimator, X_test, y_test)
    608 
    609     error_msg = ("scoring must return a number, got %s (%s) "

~\anaconda3\lib\site-packages\sklearn\metrics\_scorer.py in __call__(self, estimator, *args, **kwargs)
     86             if isinstance(scorer, _BaseScorer):
     87                 score = scorer._score(cached_call, estimator,
---> 88                                       *args, **kwargs)
     89             else:
     90                 score = scorer(estimator, *args, **kwargs)

~\anaconda3\lib\site-packages\sklearn\metrics\_scorer.py in _score(self, method_caller, clf, X, y, sample_weight)
    299         y_type = type_of_target(y)
    300         if y_type not in ("binary", "multilabel-indicator"):
--> 301             raise ValueError("{0} format is not supported".format(y_type))
    302 
    303         if is_regressor(clf):

ValueError: multiclass format is not supported

**Comment by solegalli** [comment]
Did you try changing the scoring metric when you set up the tree discretiser? instead of roc_auc maybe try accuracy? does that help?

The problem is not in feature-engine, is in sklearn, and I think it might be related to the metric.

**Comment by ashok49473** [comment]
> Hi @ashok49473
> 
> Did you try changing the scoring metric when you set up the tree discretiser? instead of roc_auc maybe try accuracy? does that help?
> 
> The problem is not in feature-engine, is in sklearn, and I think it might be related to the metric.
